flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/physical/stream/ExpandWindowTableFunctionTransposeRule.scala

```diff
@@ -0,0 +1,298 @@
+import org.apache.calcite.rel.`type`.RelDataType
+import org.apache.calcite.rex.{RexInputRef, RexLiteral, RexNode, RexProgram}
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * This rule transposes [[StreamPhysicalExpand]] past [[StreamPhysicalWindowTableFunction]] to make
+ * [[PushWindowTableFunctionIntoWindowAggregateRule]] can match the rel tree pattern and optimize
+ * them into [[StreamPhysicalWindowAggregate]].
+ *
+ * Example:
+ *
+ * MyTable: a INT, c STRING, rowtime TIMESTAMP(3)
+ *
+ * SQL:
+ * {{{
+ * SELECT
+ *    window_start,
+ *    window_end,
+ *    count(distinct a),
+ *    count(distinct c)
+ * FROM TABLE(TUMBLE(TABLE MyTable, DESCRIPTOR(rowtime), INTERVAL '15' MINUTE))
+ * GROUP BY window_start, window_end
+ * }}}
+ *
+ * We will get part of the initial physical plan like following:
+ * {{{
+ * WindowAggregate(groupBy=[$f4, $f5], window=[TUMBLE(win_start=[window_start],
+ * win_end=[window_end], size=[15 min])], select=[$f4, $f5, COUNT(DISTINCT a) FILTER $g_1 AS $f2,
+ * COUNT(DISTINCT c) FILTER $g_2 AS $f3, start('w$) AS window_start, end('w$) AS window_end])
+ * +- Exchange(distribution=[hash[$f4, $f5]])
+ *    +- Calc(select=[window_start, window_end, a, c, $f4, $f5, =($e, 1) AS $g_1, =($e, 2) AS $g_2])
+ *       +- Expand(projects=[{window_start, window_end, a, c, $f4, null AS $f5, 1 AS $e},
+ *       {window_start, window_end, a, c, null AS $f4, $f5, 2 AS $e}])
+ *          +- Calc(select=[window_start, window_end, a, c,
+ *          MOD(HASH_CODE(a), 1024) AS $f4, MOD(HASH_CODE(c), 1024) AS $f5])
+ *             +- WindowTableFunction(window=[TUMBLE(time_col=[rowtime], size=[15 min])])
+ * }}}
+ *
+ * However, it can't match [[PushWindowTableFunctionIntoWindowAggregateRule]], because
+ * [[StreamPhysicalWindowTableFunction]] is not near [[StreamPhysicalWindowAggregate]].
+ * So we need to transpose [[StreamPhysicalExpand]] past [[StreamPhysicalWindowTableFunction]]
+ * to make the part of rel tree like this which can be matched by
+ * [[PushWindowTableFunctionIntoWindowAggregateRule]].
+ *
+ * {{{
+ * WindowAggregate(groupBy=[$f4, $f5], window=[TUMBLE(win_start=[window_start],
+ * win_end=[window_end], size=[15 min])], select=[$f4, $f5, COUNT(DISTINCT a) FILTER $g_1 AS $f2,
+ * COUNT(DISTINCT c) FILTER $g_2 AS $f3, start('w$) AS window_start, end('w$) AS window_end])
+ * +- Exchange(distribution=[hash[$f4, $f5]])
+ *   +- Calc(select=[window_start, window_end, a, c, $f4, $f5, ($e = 1) AS $g_1, ($e = 2) AS $g_2])
+ *     +- WindowTableFunction(window=[TUMBLE(time_col=[rowtime], size=[15 min])])
+ *       +- Expand(...)
+ * }}}
+ * </pre>
+ */
+class ExpandWindowTableFunctionTransposeRule
+  extends RelOptRule(
+    operand(classOf[StreamPhysicalExpand],
+      operand(classOf[StreamPhysicalCalc],
+        operand(classOf[StreamPhysicalWindowTableFunction], any()))),
+    "ExpandWindowTableFunctionTransposeRule") {
```

Comment:
@wuchong Could we update the rule to match `WindowAggregate` -> `Exchange` -> `Calc` (Optional) -> Expand -> Calc -> WindowTableFunction, and directly convert to equivalent tree which merge `WindowTableFunction` into `WindowAggregate`.